common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,142 @@
+                if (this.endTime) {
+                    this.completeAfter = this.calculateCompleteAfter(this.startTime, this.endTime);
+                }
+                if (this.isEnabled) {
+                    this.bindHandlers();
+                }
+            },
+
+        /** Bind any necessary function callbacks to DOM events. */
+            bindHandlers: function() {
+                var self = this;
+
+            /** Event handler to check if the video is complete, and submit a completion if it is.
+             *
+             *  If the timeupdate handler doesn't fire after the required percentage, this will catch
+             *  any fully complete videos.
+             **/
+                this.state.el.on('ended.completion', function() {
+                    self.handleEnded();
+                });
+
+            /** Event handler to check video progress, and mark complete if
+             *  greater than completionPercentage
+             **/
+                this.state.el.on('timeupdate.completion', function(ev, currentTime) {
+                    self.handleCompletion(currentTime);
+                });
+            },
+
+            handleEnded: function() {
+                if (this.complete) {
+                    return;
+                }
+                this.markCompletion();
+            },
+
+            handleCompletion: function(currentTime) {
+                var duration;
+                if (this.complete) {
+                    return;
+                }
+                if (this.lastSentAt !== undefined && currentTime - this.lastSentAt < this.repostDelay()) {
+                    // Throttle attempts to submit in case of network issues
+                    return;
+                }
+                if (this.completeAfter === undefined) {
+                    // Duration is not available at initialization time
+                    duration = this.state.videoPlayer.duration();
+                    if (!duration) {
+                        // duration is not yet set. Wait for another event,
+                        // or fall back to 'ended' handler.
+                        return;
+                    }
+                    this.completeAfter = this.calculateCompleteAfter(this.startTime, duration);
+                }
+
+                if (currentTime > this.completeAfter) {
+                    this.lastSentAt = currentTime;
+                    this.markCompletion();
+                }
+            },
+
+            markCompletion: function() {
+                var self = this;
+                var errmsg;
+                this.complete = true;
+                if (this.state.config.publishCompletionUrl) {
+                    $.ajax({
+                        type: 'POST',
+                        url: this.state.config.publishCompletionUrl,
+                        contentType: 'application/json',
+                        dataType: 'json',
+                        data: JSON.stringify({completion: 1.0}),
```

Comment:
@jcdyer I messed around with a couple of the [jquery.ajax options for posting data](https://api.jquery.com/jQuery.ajax/#sending-data-to-server), like `processData: false` and different options for `contentType`, but had no luck getting the data parsed properly by the `XBlock.json_handler`.

Also found this [stackoverflow post](https://stackoverflow.com/a/6323528/4302112) which suggests stringifying the data like you did.

So no worries here :)